code/modules/mining/boulder_processing/boulder.dm

```diff
@@ -0,0 +1,262 @@
+		playsound(loc, 'sound/effects/picaxe1.ogg', 60, FALSE)
+	qdel(src)
+
+/**
+ * This is called when a boulder is spawned from a vent, and is used to set the boulder's icon as well as durability.
+ * We also set our boulder_size variable, which is used for inheiriting the icon_state later on if processed.
+ */
+/obj/item/boulder/proc/flavor_boulder(obj/structure/ore_vent/parent_vent, size = BOULDER_SIZE_SMALL, is_artifact = FALSE)
+	var/durability_min = size
+	var/durability_max = size + BOULDER_SIZE_SMALL
+	if(parent_vent)
+		durability_min = parent_vent.boulder_size
+		durability_max = parent_vent.boulder_size + BOULDER_SIZE_SMALL
+	durability = rand(durability_min, durability_max) //randomize durability a bit for some flavor.
+	boulder_size = size
+	if(parent_vent)
+		boulder_size = parent_vent.boulder_size
+		boulder_string = parent_vent.boulder_icon_state
+	update_appearance(UPDATE_ICON_STATE)
+
+/obj/item/boulder/artifact
+	name = "artifact boulder"
+	desc = "This boulder is brimming with strange energy. Cracking it open could contain something unusual for science."
+	icon_state = "boulder_artifact"
+	/// This is the type of item that will be inside the boulder. Default is a strange object.
+	var/artifact_type = /obj/item/relic
+	/// References to the relic inside the boulder, if any.
+	var/obj/item/artifact_inside
+
+/obj/item/boulder/artifact/Initialize(mapload)
+	. = ..()
+	artifact_inside = new artifact_type(src) /// This could be poggers for archaeology in the future.
+
+/obj/item/boulder/artifact/Destroy(force)
+	artifact_inside = null
+	return ..()
+
+/obj/item/boulder/artifact/convert_to_ore()
+	. = ..()
+	artifact_inside.forceMove(drop_location())
+	artifact_inside = null
+
+/obj/item/boulder/artifact/break_apart()
+	artifact_inside = null
+	return ..()
+
+/obj/item/boulder/artifact/update_icon_state()
+	. = ..()
+	icon_state = "boulder_artifact" //We're always going to use the artifact state for clarity.
+	return
+
+/obj/item/boulder/gulag
+	name = "boulder"
+	desc = "This rocks. It's also a gulag boulder, so it's probably not worth as much."
+	var/list/pick_minerals = list(
+		/datum/material/diamond = 1,
+		/datum/material/gold = 8,
+		/datum/material/iron = 95,
+		/datum/material/plasma = 30,
+		/datum/material/silver = 20,
+		/datum/material/titanium = 8,
+		/datum/material/uranium = 3,
+	)
+
+/obj/item/boulder/gulag/Initialize(mapload)
+	. = ..()
+	add_gulag_minerals()
+
+/**
+ * Unique proc for gulag boulders, which adds a random amount of minerals to the boulder.
+ */
+/obj/item/boulder/gulag/proc/add_gulag_minerals()
+	var/datum/material/new_material = pick_weight(pick_minerals)
+	var/list/new_mats = list()
+	new_mats += new_material
+	new_mats[new_material] = SHEET_MATERIAL_AMOUNT * rand(1,3) //We only want a few sheets of material in the gulag boulders
+	set_custom_materials(new_mats)
+
+/obj/item/boulder/gulag/volcanic
+	pick_minerals = list(
```

Comment:
One good solution would be to first create a static list containing all your minerals(say `volcanic_minerals`) and then assign that to pick minerals i.e. `pick_minerals = volcanic_minerals`. Since there is goanna be a ton of boulders generated in game this can save a lot of memory.